build-tools/xaprepare/xaprepare/Steps/Step_GenerateCGManifest.cs

```diff
@@ -0,0 +1,302 @@
+					if (pi < 1 || e [pi - 1] != ' ')
+						continue;
+					hash    = e.Substring (1, pi - 2);
+					break;
+				}
+				return new GitSubmoduleInfo () {
+					RepositoryUrl   = url,
+					CommitHash      = hash,
+				};
+			}
+		}
+
+		static string? GetSubmoduleId (string line)
+		{
+			int eq = line.IndexOf ('=');
+			if (eq < 0)
+				return null;
+			int lastDot = line.LastIndexOf ('.', eq);
+			return line.Substring (Submodule.Length, lastDot - Submodule.Length);
+		}
+	}
+
+	sealed class MSBuildPackageReferenceInfo : CGManifestEntry {
+
+		string name;
+
+		public  override    string      Name    => name;
+		public  string      Version {get; private set;}
+
+		private MSBuildPackageReferenceInfo (string name, string version)
+		{
+			this.name = name;
+			Version = version;
+		}
+
+		public override void WriteJsonTo (StreamWriter writer, HashSet<string> developmentDependencies)
+		{
+			string dev = developmentDependencies.Contains (Name)
+				? "true"
+				: "false";
+
+			writer.WriteLine ($"        {{");
+			writer.WriteLine ($"            \"Component\": {{");
+			writer.WriteLine ($"                \"Type\": \"nuget\",");
+			writer.WriteLine ($"                \"Nuget\": {{");
+			writer.WriteLine ($"                    \"Name\": \"{Name}\",");
+			writer.WriteLine ($"                    \"Version\": \"{Version}\"");
+			writer.WriteLine ($"                }}");
+			writer.WriteLine ($"            }},");
+			writer.WriteLine ($"            \"DevelopmentDependency\":{dev}");
+			writer.Write     ($"        }}");
+		}
+
+		static  readonly    XNamespace      MSBuildXmlns    = XNamespace.Get ("http://schemas.microsoft.com/developer/msbuild/2003");
+
+		public static IEnumerable<MSBuildPackageReferenceInfo> GetPackageReferences ()
+		{
+			var files = Directory.EnumerateFiles (BuildPaths.XamarinAndroidSourceRoot, "*.csproj", SearchOption.AllDirectories)
+				.Concat (Directory.EnumerateFiles (BuildPaths.XamarinAndroidSourceRoot, "*.targets", SearchOption.AllDirectories))
+				.Concat (Directory.EnumerateFiles (BuildPaths.XamarinAndroidSourceRoot, "*.projitems", SearchOption.AllDirectories))
+				;
+			var packages    = new Dictionary<string, HashSet<string>> ();
+			var versions    = new Dictionary<string, HashSet<string>> ();
+			foreach (var file in files) {
+				var contents            = File.ReadAllText (file);
+				if (contents.IndexOf ("PackageReference", StringComparison.Ordinal) < 0 ||
+						contents.IndexOf ("PropertyGroup", StringComparison.Ordinal) < 0)
+					continue;
+				var proj                = XDocument.Parse (contents);
+				var packageReferences   = proj.Elements (MSBuildXmlns + "Project")
+					.Elements (MSBuildXmlns + "ItemGroup")
+					.Elements (MSBuildXmlns + "PackageReference");
+				foreach (var packageReference in packageReferences) {
+					var name    = (string) packageReference.Attribute ("Include");
+					var version = (string) packageReference.Attribute ("Version") ??
```

Comment:
Shouldn't the casts here and in the previous line be to `string?`? Line 265 below makes a comparison against `null` and `string` cannot be `null` with NRT.

Reply:
Indeed!